firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java

```diff
@@ -581,12 +581,19 @@ QueryData getQueryData(Target target) {
     int targetId;
   }
 
-  /** Unpin all the documents associated with the given target. */
+  /**
+   * Unpin all the documents associated with the given target.
+   *
+   * <p>Releasing a non-existing target is a no-op.
```

Comment:
Hrm...  why do we allow this now?  It seems like we should not be double-releasing targets (if we do, it suggests we released it too early). But I may be confused or missing some context.

Reply:
Sorry, I should have included more context in the PR.

We decided to do this (make LocalStore.allocateTarget|releaseTarget and RemoteStore.listen|unlisten) mainly because multi-tab in JS make it really difficult to track the state of targets in local store and remote store. The amount of code I need to put into SyncEngine to track each target to avoid double-listen or double-allocate/release made it hard to understand.

I suppose with Android and iOS we could have avoided this and make sure sync engine never double listen. I still do it  to mainly keep consistent with JS.

But I also feel like the old way was kind of leaking implementation details upwards,  in the sense that sync engine must be aware and check for double listen for local/remote store. The fact that 1:N target query mapping + multi-tab made this really obvious and painful. So maybe it's not a bad idea to make all of those method re-entrant?

Reply:
Thanks for the explanation.

I'm not sure I want to push back too hard, but I do want to clarify my understanding.  It seems like the calling code still needs to do the appropriate tracking / accounting in order to avoid double-calling unlisten or releaseTarget (I'm less concerned about double-calling listen or allocateTarget).  If we double-call unlisten, doesn't that mean we likely unlistened too early and therefore potentially missed some server updates for the target?  Similarly, double-calling releaseTarget likely means we released the target too early, potentially making it eligible for garbage collection when it shouldn't be.

Or am I confused and there are cases when it's normal / safe for the SDK code to e.g. unlisten to the same target multiple times?

In my mind the tracking / accounting we were doing was unavoidable and so we made listen/unlisten strict about it just to sanity-check that we were doing it right.

Given this is just porting the JS logic, I'll go ahead and approve for now, but maybe we can sync up tomorrow at standup?

Reply:
And by "tomorrow at standup" I mean Monday or whenever you're back in the office. 😂

Reply:
FYI- I tried adding the strict checking back to unlisten() and releaseTarget() in the JS SDK and all the tests still seem to pass (which is a relief since I still think if they didn't pass that would indicate a bug).  I'm not concerned about listen() and allocateTarget().  I think it's relatively safe to allow calling them repeatedly.

So I think the question is whether we think it's valuable to prevent double calling unlisten() and releaseTarget() (and whether it's an okay contract to allow double-listen but not double-unlisten).  Let's sync up when you're back!